### PR TITLE
Bucket CNAMES and Pages

### PR DESCRIPTION
**Pull request by sanderpick**
- Creates a CNAME for each (global) bucket
- The gateway leverages the client to implement a simple _bucket file system_ for rendering static assets via a static middleware

### CODE REVIEW

File: api/client/client.go [hunk 1]
```diff
@@ -275,7 +275,9 @@ func (c *Client) PushBucketPath(
 				Payload: &pb.PushBucketPathRequest_Chunk{
 					Chunk: buf[:n],
 				},
-			}); err != nil {
+			}); err == io.EOF {
+				break
```
Comment (jsign) [inline api/client/client.go]: :sweat_smile:

File: api/service.go [hunk 2]
```diff
@@ -908,6 +914,15 @@ func (s *service) createBucket(ctx context.Context, proj *c.Project, name string
 		return nil, err
 	}
 
+	if s.dnsManager != nil {
+		parts := strings.SplitN(s.gatewayUrl, "//", 2)
+		if len(parts) > 1 {
+			if _, err := s.dnsManager.NewCNAME(name, parts[1]); err != nil {
+				return nil, err
+			}
+		}
```
Comment (jsign) [inline api/service.go]: The `else` case seems like a malformed url, so preferred to ignore?

File: gateway/bucketfs.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+// serveBucket returns a middleware handler that serves files in a bucket.
+func serveBucket(fs serveBucketFileSystem) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		bucket, err := bucketFromHost(c.Request.Host, fs.ValidHost())
+		if err != nil {
+			return
+		}
+		if fs.Exists(bucket, c.Request.URL.Path) {
+			c.Writer.WriteHeader(http.StatusOK)
+			ctype := mime.TypeByExtension(filepath.Ext(c.Request.URL.Path))
```
Comment (sanderpick) [inline gateway/bucketfs.go]: Would be nice to use more advanced content type detection, but we'd need the client to implement a `ReadSeeker`.

File: gateway/bucketfs.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), f.timeout)
+	defer cancel()
+	return f.client.PullBucketPath(ctx, path.Join(bucket, pth), writer, f.auth)
+}
+
+func (f *bucketFileSystem) ValidHost() string {
+	return f.host
+}
+
+func bucketFromHost(host, valid string) (buck string, err error) {
```
Comment (sanderpick) [inline gateway/bucketfs.go]: This prevents bucket rendering under a different domain.

File: api/client/client.go (hunk 1 above)
Comment (sanderpick) [reply api/client/client.go]: Oh yeah haha. I can remove that when digging into the nginx-related. steam error I mentioned in Slack.

File: api/service.go (hunk 2 above)
Comment (sanderpick) [reply api/service.go]: As in, ignore the request?

Comment (jsign) [reply api/service.go]: I mean, if `s.gatewayUrl` parsing doesn't have `len(parts)>1`, the current code ignores that and doesn't register the CNAME. Is that something that should happen? I wonder if a malformed `s.gatewayUrl` can backfire somewhere else (or in the future).

Comment (sanderpick) [reply api/service.go]: Oh, I see. Makes sense. We could try to parse the URL on start, that fails -> fatal?

Comment (jsign) [reply api/service.go]: It may be an option, yeah. I'll leave if that may worth it to you :P

**Pull request merged**